lib/scholar/manifold/tsne.ex

```diff
@@ -0,0 +1,302 @@
+      {opts[:perplexity], opts[:learning_rate], opts[:num_iters], opts[:num_components],
+       opts[:exaggeration], opts[:init], opts[:metric]}
+
+    x = to_float(x)
+    {n, _dims} = Nx.shape(x)
+
+    y1 =
+      case init do
+        :random ->
+          key = Nx.Random.key(seed)
+
+          {y, _new_key} =
+            Nx.Random.normal(key, 0.0, 1.0e-4, shape: {n, num_components}, type: Nx.type(x))
+
+          y
+
+        :pca ->
+          Scholar.Decomposition.PCA.fit_transform(x, num_components: num_components)
+      end
+
+    p = p_joint(x, perplexity, metric)
+
+    {y, _, _, _, _} =
+      while {y1, y2 = y1, learning_rate, p, i = 2}, i < num_iters do
+        q = q_joint(y1, metric)
+        grad = gradient(p * exaggeration(i, exaggeration), q, y1, metric)
+        y_next = y1 - learning_rate * grad + momentum(i) * (y1 - y2)
+        {y_next, y1, learning_rate, p, i + 1}
+      end
+
+    y
+  end
+
+  defnp pairwise_dist(x, metric) do
+    {num_samples, num_features} = Nx.shape(x)
+
+    t1 =
+      x
+      |> Nx.reshape({1, num_samples, num_features})
+      |> Nx.broadcast({num_samples, num_samples, num_features})
+
+    t2 =
+      x
+      |> Nx.reshape({num_samples, 1, num_features})
+      |> Nx.broadcast({num_samples, num_samples, num_features})
+
+    case metric do
+      :squared_euclidean ->
+        Distance.squared_euclidean(t1, t2, axes: [2])
+
+      :euclidean ->
+        Distance.euclidean(t1, t2, axes: [2])
+
+      :manhattan ->
+        Distance.manhattan(t1, t2, axes: [2])
+
+      :cosine ->
+        Distance.cosine(t1, t2, axes: [2])
+
+      :chebyshev ->
+        Distance.chebyshev(t1, t2, axes: [2])
+    end
+  end
+
+  defnp p_conditional(distances, sigmas) do
+    arg = -distances / (2 * Nx.reshape(sigmas, {:auto, 1})) ** 2
+    {n, _} = Nx.shape(arg)
+
+    # Set diagonal to a large negative number so it becomes 0 after applying exp()
+    min_value = Nx.Constants.min_finite(Nx.type(arg))
+    arg_with_min_diag = Nx.put_diagonal(arg, Nx.broadcast(min_value, {n}))
+
+    stabilization_constant = Nx.reduce_max(arg_with_min_diag, axes: [1], keep_axes: true)
+    arg = arg - stabilization_constant
+
+    p = Nx.exp(arg)
+    p / Nx.sum(p, axes: [1], keep_axes: true)
+  end
+
+  defnp perplexity(condition_matrix) do
+    exponent = -Nx.sum(condition_matrix * Nx.log2(condition_matrix), axes: [1])
+    2 ** exponent
+  end
+
+  defnp find_sigmas(distances, target_perplexity) do
+    {n, _} = Nx.shape(distances)
+    sigmas = Nx.broadcast(Nx.tensor(0, type: Nx.type(distances)), {n})
+
+    {sigmas, _, _, _} =
+      while {sigmas = sigmas, distances = distances, target_perplexity, i = 0}, i < n do
+        distances_i = Nx.take(distances, i)
+
+        sigmas =
+          Nx.indexed_put(
+            sigmas,
+            Nx.reshape(i, {1, 1}),
+            Nx.new_axis(binary_search(distances_i, target_perplexity), -1)
+          )
+
+        {sigmas, distances, target_perplexity, i + 1}
+      end
```

Comment:
same suggestion about unroll with generators applies here, but it's also not a blocker